### PR TITLE
add killWord, killWordBackwards, getNextChar, getPrevChar, and getWord functions

### PR DESCRIPTION
**Pull request by dagit**
I had some spare time so I started prototyping some new editing capabilities. This pull request is to help give us something concrete while we discuss:
- Which features/capabilities we want to add
- Where they go (which package, eg., `text-zipper` vs. `brick` vs. `matterhorn`)
- misc implementation concerns

For now, I added them all in `text-zipper` because it was useful to access the `TextZipper` internals (especially for `getNextChar`/`getPrevChar`), but I think mostly these things could live outside of `text-zipper`. There are more efficient ways to implement `killWord` and `killWordBackwards`, but I like writing them on top of the other functions for maintenance reasons. The implementations in this pull request are a complete rewrite to better support the notion of a "word", currently it follows the C rules for identifier names (alpha numeric + underscore). I got this spec from a man page.

Here is the run down:
- `killWord` -- Inspired by zsh ESC-d (I say "inspired by" because I may not have faithfully mimic'd it)
- `killWordBackwards` -- Inspired by zsh M-w
- `getNextChar` -- Gets the character to Right of the cursor eg., the character that would be deleted by `deleteChar`.
- `getPrevChar` -- Gets the character to the Left of the cursor eg., the character that would be deleted by `deletePrevChar`.
- `getWord` -- Experimental thing I was trying out. A function like this would actually be very useful for tab completion within a sentence (something we current get horribly wrong in matterhorn). If the cursor is positioned on a word it attempts to fetch that word. I also added the unexported function `getThing`. My inspiration was the emacs library [thing at point](https://www.emacswiki.org/emacs/ThingAtPoint).

People who might want to join the discussion: @elliottt @jtdaugherty @aisamanra

### CODE REVIEW

File: src/Data/Text/Zipper.hs [hunk 1]
```diff
@@ -247,6 +253,69 @@ killToBOL :: Monoid a => TextZipper a -> TextZipper a
...
+
+getThing :: Monoid a
+         => (TextZipper a -> Maybe a) -- ^ get characters to the left
+         -> (TextZipper a -> Maybe a) -- ^ get characters to the right
+         -> TextZipper a
+         -> Maybe a
+getThing getLeft getRight tz =
+  combine (getLeft tz) (getRight tz)
+  where
+  -- XXX: Does this exist in the libraries already?
```
Comment (jtdaugherty) [inline src/Data/Text/Zipper.hs:301]: Yay, looks like it does:

```
Prelude Data.Monoid> Just [5] <> Just [6]
Just [5,6]
Prelude Data.Monoid> Just [5] <> Nothing
Just [5]
```

**Comment by jtdaugherty** [comment]
This is great, thanks! Can you add haddock for all exported functions? Then I'll merge!

**Comment by jtdaugherty** [comment]
Hey Jason, do you have time to add the haddock to this patch?

**Comment by hpdeifel** [comment]
This looks very similar to what I implemented over at https://github.com/hpdeifel/hledger-iadd/blob/master/src/Data/Text/Zipper/Generic/Words.hs.

A few thoughts:
  - I also implemented `moveWord`/`moveWordLeft`, since they rely on the same underlying principle as `killWord`/`killWordBackwards`.
  - The implementation of "movement by words" differs between emacs, bash/readline and zsh. I would prefer bash/readline, because more people are familiar with it.
  - Special care has to be taken with multi-line zippers. I didn't test what this pull request does on line boundaries, but they should also be regarded as word boundaries. It would also be nice if `getNextChar`/`getPrevChar` had a way to distinguish newlines from EOF/BOF and other characters.

I wrote tests for my implementation: https://github.com/hpdeifel/hledger-iadd/blob/master/tests/Data/Text/Zipper/Generic/WordsSpec.hs. They can be read as a specification of the semantics I had in mind.

**Comment by jtdaugherty** [comment]
@hpdeifel I'd like to integrate your implementation and its test suite. I'll take care of that process, though.

@dagit, thank you for your work on this!

**Pull request closed without merging**